src/diff/mod.rs

```diff
@@ -1,36 +1,278 @@
+        -> Result<(), String> {
+
+        // TODO: Consider storing directory contents in sorted order
+        let old = get_sorted_contents(old);
+        let new = get_sorted_contents(new);
+
+        let mut old_iter = old.iter();
+        let mut new_iter = new.iter();
+        let mut old_entry_opt = old_iter.next();
+        let mut new_entry_opt = new_iter.next();
+
+        while old_entry_opt.is_some() || new_entry_opt.is_some() {
+            // need to skip Repos
+            while let Some(DirectoryContents::Repo) = old_entry_opt {
+                old_entry_opt = old_iter.next();
+            }
+            while let Some(DirectoryContents::Repo) = new_entry_opt {
+                new_entry_opt = new_iter.next();
+            }
+            match (old_entry_opt, new_entry_opt) {
+                (Some(old_entry), Some(new_entry)) => {
+                    let cmp = new_entry.label().cmp(&old_entry.label());
+                    match cmp {
+                        Ordering::Greater => {
+                            diff.add_deleted_files(old_entry, parent_path)?;
+                            old_entry_opt = old_iter.next();
+                        },
+                        Ordering::Less => {
+                            diff.add_created_files(new_entry, parent_path)?;
+                            new_entry_opt = new_iter.next();
+                        },
+                        Ordering::Equal => {
+                            use DirectoryContents::{File, SubDirectory};
+                            match (new_entry, old_entry) {
+                                (File(new_file), File(old_file)) => {
+                                    if old_file.size != new_file.size || &old_file.checksum() != &new_file.checksum() {
+                                        diff.add_modified_file(new_file, parent_path);
+                                    }
+                                },
+                                (File(new_file), SubDirectory(old_dir)) => {
+                                    diff.add_created_file(new_file, parent_path);
+                                    diff.add_deleted_files(old_entry, parent_path)?;
+                                },
+                                (SubDirectory(new_dir), File(old_file)) => {
+                                    diff.add_created_files(new_entry, parent_path)?;
+                                    diff.add_deleted_file(old_file, parent_path);
+                                },
+                                (SubDirectory(new_dir), SubDirectory(old_dir)) => {
+                                    parent_path.borrow_mut().push(new_dir.label.clone());
+                                    Diff::collect_diff(&**old_dir, &**new_dir, parent_path, diff)?;
+                                    parent_path.borrow_mut().pop();
+                                },
+                                _ => panic!("should not happen unless the algo is incorrect")
```

Comment:
Can we add the check for `Repo` here instead of above? This would avoid the panic and read well for covering all cases. Also it avoids the fact that a case might be added and then suddenly things panic by accident.

Reply:
Outdated, will respond in the newer comment